### PR TITLE
better-emojis: refactor

### PR DESCRIPTION
**Pull request by Weredime**
Refactoring better-emojis to have:

- [x] better code
- [x] dynamic enable/disable

### CODE REVIEW

File: addons/better-emojis/addon.json [hunk 1]
```diff
@@ -15,9 +15,11 @@
   "userscripts": [
     {
       "url": "userscript.js",
-      "matches": ["https://scratch.mit.edu/*"]
+      "matches": ["projects", "studios", "https://scratch.mit.edu/users/*/"]
```
Comment (apple502j) [inline addons/better-emojis/addon.json:18]: What happens to message pages?

Comment (Weredime) [reply addons/better-emojis/addon.json]: Fixed

File: addons/better-emojis/addon.json [hunk 2]
```diff
@@ -15,9 +15,11 @@
   "userscripts": [
     {
       "url": "userscript.js",
-      "matches": ["https://scratch.mit.edu/*"]
+      "matches": ["projects", "studios", "https://scratch.mit.edu/users/*/", "https://scratch.mit.edu/messages"]
```
Comment (Samq64) [inline addons/better-emojis/addon.json:18]:
```suggestion
      "matches": ["projects", "studios", "profiles", "https://scratch.mit.edu/messages"]
```

Comment (Weredime) [reply addons/better-emojis/addon.json]: Done

File: addons/better-emojis/userscript.js [hunk 3]
```diff
@@ -1,95 +1,53 @@
...
+  addon.self.addEventListener("reenabled", () => {
+    const emojis = document.querySelectorAll("img.easter-egg[data-emoji], img.emoji[data-emoji]");
+    for (const emoji of emojis) {
+      emoji.src = addon.self.dir + "/images/" + emoji.dataset.emoji + ".svg";
+    }
+  });
+  while (true) {
+    const emoji = await addon.tab.waitForElement(".emoji, .easter-egg", {
+      markAsSeen: true,
+      reduxCondition: (state) => state?.scratchGui.mode.isPlayerOnly,
```
Comment (apple502j) [inline addons/better-emojis/userscript.js:41]: Won't this crash in studios?

**Review by apple502j — APPROVED** [review]

**Review by apple502j — CHANGES_REQUESTED** [review]
Wait, pressed the wrong button

**Review by apple502j — APPROVED** [review]

File: addons/better-emojis/userscript.js [hunk 4]
```diff
@@ -1,95 +1,60 @@
...
-  updateEmojis();
-  const obs = new MutationObserver(updateEmojis);
-  obs.observe(document.body, {
-    childList: true,
-    subtree: true,
   });
+  while (true) {
+    const emoji = await addon.tab.waitForElement(".emoji, .easter-egg", {
+      markAsSeen: true,
+      reduxCondition: (state) => state?.comments.status.comments === "FETCHED",
```
Comment (WorldLanguages) [inline addons/better-emojis/userscript.js:47]: `state` is guaranteed to be an object.

File: addons/better-emojis/userscript.js [hunk 5]
```diff
@@ -1,95 +1,60 @@
...
-      case "https://scratch.mit.edu/images/emoji/waffle.png":
-        el.src = addon.self.dir + "/images/waffle.svg";
-        break;
+    const match = emoji.src.match(COMMENTS_REGEX);
+    if (match !== null) {
+      const emojiId = match[addon.tab.clientVersion === "scratch-www" ? 1 : 2];
+      if (EMOJIS.includes(emojiId)) {
+        // check if emoji is supported by this addon
+        if (!addon.self.disabled) emoji.src = addon.self.dir + "/images/" + emojiId + ".svg";
+        emoji.dataset.emoji = emojiId;
```
Comment (WorldLanguages) [inline addons/better-emojis/userscript.js:56]: Can we use `.dataset.saEmoji`?

**Comment by WorldLanguages** [comment]
I wasn't aware this addon, when enabled, has always used a MutationObserver on the whole body on every page :S
Nice idea to refactor it!

File: addons/better-emojis/userscript.js (hunk 4 above)
Comment (WorldLanguages) [reply addons/better-emojis/userscript.js:47]: I think you meant the opposite?
`state.comments?.status?.comments`
Current code in your branch crashes on the messages page

**Review by WorldLanguages — CHANGES_REQUESTED** [review]
Doesn't work in the messages page

**Comment by Weredime** [comment]
@WorldLanguages I've addressed your changes

**Review by WorldLanguages — APPROVED** [review]

**Pull request merged**